Make feather stores read incrementally (#2805)
tenzir/vast: web/blog/vast-v2.4.1/index.md

---
title: VAST v2.4.1

> description: Faster Initial Results

Reviewer: ```suggestion
description: Faster Query Taste
```

or "Faster Time to Taste"

authors: dominiklohmann
date: 2022-12-19
tags: [release, feather, performance]
---

[VAST v2.4.1][github-vast-release] improves the performance of queries when VAST
is under high load, and significantly reduces the time to first result for
queries with a low selectivity.

[github-vast-release]: https://github.com/tenzir/vast/releases/tag/v2.4.1

<!--truncate-->

## Reading Feather Files Incrementally

VAST's Feather store naïvely used the [Feather reader][feather-reader] from the
Apache Arrow C++ library in its initial implementation. However, its API is

> rather limited: It does not support reading incrementally. We've swapped this

Reviewer: ```suggestion
rather limited: It does not support reading record batches incrementally. We've swapped this
```

I was stuck with the "what are we reading" when going over this.

out with a more efficient implementation that does.

[feather-reader]: https://github.com/apache/arrow/blob/apache-arrow-10.0.1/cpp/src/arrow/ipc/feather.h#L57-L108

This is best explained visually:

![Incremental Reads](incremental-reads-light.png#gh-light-mode-only)
![Incremental Reads](incremental-reads-dark.png#gh-dark-mode-only)

> Within the scope of a single store, queries take the same amount of time

Author: I reworded that to say "Feather store file," which picks up what we had in the paragraph before.

> overall, but there exist two distinct advantages of this approach:

Reviewer: Is it really the same performance? The drawing suggests that I/O and compute are now interleaved, which I would imagine is faster due to pipelining.

Author: I made it more explicit that this is about a single query in a single store.

1. The first result arrives much faster at the client.
2. Stores do less work for cancelled queries.

> One additional benefit that is not immediately obvious only comes into play a

Reviewer: ```suggestion
One additional benefit that is not immediately obvious comes into play when a
```

> query arrives at multiple stores in parallel. Because the dksk reads are more

Reviewer: ```suggestion
query arrives at multiple stores in parallel: disk reads are more
```

> evenly spread out now, the disk reads are less likely to overlap between stores

Reviewer: ```suggestion
evenly spread out now, making them less likely to overlap between stores.
```

> queries in parallel. For disk-read performance-bound deployments this can lead

Reviewer: ```suggestion
For deployments wither slower I/O paths this can lead
```

to a significant query performance improvement.

> To verify and test this, I've created a VAST database with 300M Zeek events from

Reviewer: ```suggestion
To verify and test this, we've created a VAST database with 300M Zeek events.
```

Events alone is not sufficient to express the juiciness of a dataset. You need to add the actual size in GiB.

Author: I've added this in parantheses.

> a Corelight sensor. All tests were performed on a cold start of VAST, i.e., I

Reviewer: ```suggestion
All tests were performed on a cold start of VAST, i.e., we
```

stopped and started VAST after every repetition of each test.

> I performed three tests:

Reviewer: ```suggestion
We performed three tests:

```

> 1. Export a single event (20 repetitions)
> 2. Export all events (20 repetitions)
> 3. Rebuild the entire database (3 repetitions)

Reviewer: ```suggestion
1. Export a single event (20 times)
2. Export all events (20 times)
3. Rebuild the entire database (3 times)
```

Reviewer: Can you add a linke to rebuild?
